From 47332b75b41734dbc331b98d7d7bca6c987b4406 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 16 Jan 2024 14:38:45 +0100
Subject: [PATCH 0517/1695] drivers: clk: fix RCC pinctrl

Load pinctrl state at driver initialization and at PM resume.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c3c292680bb632965577f295d0b2ee5b2460c4b4 drivers: clk: fix RCC pinctrl]
Change-Id: Id056eefca342f05c9d0b7155340d45d3eb2a133e
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/351823
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/clk/clk-stm32-rcc.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/core/drivers/clk/clk-stm32-rcc.c b/core/drivers/clk/clk-stm32-rcc.c
index 51d0da41c..e41622d78 100644
--- a/core/drivers/clk/clk-stm32-rcc.c
+++ b/core/drivers/clk/clk-stm32-rcc.c
@@ -21,6 +21,7 @@ struct stm32mp_rcc_clock {
 
 static SLIST_HEAD(, stm32mp_rcc_clock) stm32mp_rcc_list =
 	SLIST_HEAD_INITIALIZER(stm32mp_rcc_list);
+static struct pinctrl_state *rcc_pinctrl;
 
 static TEE_Result
 stm32mp_rcc_pm(enum pm_op op, unsigned int pm_hint __unused,
@@ -28,6 +29,10 @@ stm32mp_rcc_pm(enum pm_op op, unsigned int pm_hint __unused,
 {
 	struct stm32mp_rcc_clock *rcc_clock = NULL;
 
+	if (op == PM_OP_RESUME && rcc_pinctrl &&
+	    pinctrl_apply_state(rcc_pinctrl))
+		panic();
+
 	SLIST_FOREACH(rcc_clock, &stm32mp_rcc_list, link) {
 		if (op == PM_OP_RESUME) {
 			if (clk_enable(rcc_clock->clk)) {
@@ -46,21 +51,24 @@ stm32mp_rcc_pm(enum pm_op op, unsigned int pm_hint __unused,
 static TEE_Result stm32mp_rcc_probe(const void *fdt, int node,
 				    const void *compat_data __unused)
 {
-	struct pinctrl_state *pinctrl_cfg = NULL;
 	TEE_Result res = TEE_SUCCESS;
 	struct clk *clk = NULL;
 	int idx = 0;
 
 	/* Apply pinctrl configuration if present */
-	res = pinctrl_get_state_by_idx(fdt, node, 0, &pinctrl_cfg);
+	res = pinctrl_get_state_by_idx(fdt, node, 0, &rcc_pinctrl);
 	if (res && res != TEE_ERROR_ITEM_NOT_FOUND) {
 		if (res == TEE_ERROR_DEFER_DRIVER_INIT)
 			return res;
 		panic("Failed to get RCC pinctrl");
 	}
 
-	if (pinctrl_cfg)
-		stm32_pinctrl_set_secure_cfg(pinctrl_cfg, true);
+	if (rcc_pinctrl) {
+		stm32_pinctrl_set_secure_cfg(rcc_pinctrl, true);
+		res = pinctrl_apply_state(rcc_pinctrl);
+		if (res)
+			return res;
+	}
 
 	/* Enable clock(s) if present */
 	while (!clk_dt_get_by_index(fdt, node, idx++, &clk)) {
-- 
2.47.2

